### ISSUE TITLE
Weird issue with job logs

### ISSUE THREAD

**Issue by mpatel-cpi** [jhuckaby/Cronicle#26]
## Summary

First of all thanks for creating this amazing tool. We have been using it for a few weeks and absolutely love it. Great work !!!

In our QA setup everything was working fine, but when we started rolling out to production, we started seeing some issues with job logs. Only difference in our setup is Node JS version. Wondering if it is related to that?

When I schedule the job to run on the cronicle slave server, the job runs fine, but gets stuck in the end (I think during job log file cleanup):

![image](https://cloud.githubusercontent.com/assets/23645479/25592541/0761517a-2e6e-11e7-9452-e476722569de.png)

Navigating back to "Complete" jobs tab, I see "(No log file found.)" and clicking on "Download Log" or "View Full Log" shows the same.

![image](https://cloud.githubusercontent.com/assets/23645479/25592612/3f947a40-2e6e-11e7-9b36-6f1f2446905f.png)

When scheduling the same job on cronicle master, the job finishes, but the "Job Details" is blank for job "Job Event Log". Clicking "Download Log" or "View Full Log" shows the complete job log correctly.

![image](https://cloud.githubusercontent.com/assets/23645479/25592838/1d0df75c-2e6f-11e7-87d0-02303fbeca26.png)

## Steps to reproduce the problem

1. Create a new test event using Shell plugin with:
#!/bin/sh

# Enter your shell script code here
echo "Starting"
echo $CRONICLE
date
sleep 20
echo "Done."

2. Use "Run now" feature to schedule this on master or slave server and see the above mentioned issues with job logs.

## Your Setup

We have 2 separate setups for Cronicle - QA and Prod. Only difference is Node.js version, everything else is same.

### Operating system and version?

Ubuntu 16.04.2 LTS

### Node.js version?

Prod: v6.10.2
QA: v7.8.0

### Cronicle software version?

0.6.11

### Are you using a multi-server setup, or just a single server?

1 master + 1 slave

### Are you using the filesystem as back-end storage, or S3/Couchbase?

Filesystem

### Can you reproduce the crash consistently?

Yes

## Log Excerpts

Error.Log:
Failed to fetch job log file: https://172.17.2.143:3012/api/app/fetch_delete_job_log?path=%2Fhome%2Fdeployer%2Fcronicle%2FCronicle-0.6.11%2Flogs%2Fjobs%2Fjj26j9mdn02.log&auth=removed: Error: Error: Hostname/IP doesn't match certificate's altnames: "IP: 172.17.2.143 is not in the cert's list: "

**Comment by jhuckaby** [comment]
Hello there!

Thank you very much for using Cronicle, and sending in this detailed bug report.  I think I see the problem.  It looks like you are using HTTPS and also "forcing" HTTPS mode by setting both of these properties in the `WebServer` configuration section:

```js
"https": true,
"https_force": true
```

I have to be honest -- I have **never** tested either of these settings with Cronicle.  Like, not even once.  I know they work in the underlying web server component, but not with Cronicle as a whole, and especially not in a multi-server setup, where the Cronicle servers have to communicate with each other over HTTPS.  Yikes!  😮 

Unfortunately, I expect there to be many issues here, as I haven't thought about all the various places that may be affected with everything being HTTPS, and *forced* to only serve over HTTPS via `https_force`.  This is likely going to be a big can of worms.  However, I can help you with the immediate problem here:

```
Failed to fetch job log file: https://172.17.2.143:3012/api/app/fetch_delete_job_log?path=%2Fhome%2Fdeployer%2Fcronicle%2FCronicle-0.6.11%2Flogs%2Fjobs%2Fjj26j9mdn02.log&auth=removed: Error: Error: Hostname/IP doesn't match certificate's altnames: "IP: 172.17.2.143 is not in the cert's list: "
```

**Hostname/IP doesn't match certificate's altnames: "IP: 172.17.2.143 is not in the cert's list:**

Okay, so the issue here is that two Cronicle servers are trying to talk to each over via IP address to send a log file between them (which happens at the end of each job), but they are being "forced" to use HTTPS via the `https_force` property in your configuration.  The problem is, HTTPS doesn't work with IP addresses in the URL like this, unless somehow you have a special SSL Certificate that declares the IP address as part of its domain.

I think you have two options here.  First, disable `https_force` in your configuration file on all your servers.  This will still support HTTPS, but it will also "allow" plain HTTP and use that for internal communication between servers.  That should fix the log transfer issue (although there may be other issues lurking).

Second, you can upgrade to Cronicle [v0.6.14](https://github.com/jhuckaby/Cronicle/releases/tag/v0.6.14) which supports the new [web_hook_ssl_cert_bypass](https://github.com/jhuckaby/Cronicle#web_hook_ssl_cert_bypass) configuration.  Ironically, this property was only introduced for HTTPS Web Hook URLs, but as a totally accidental side effect, it should also cause server-to-server HTTP requests to support invalid SSL certificates, such as when making a HTTP/SSL call to a server via IP address.

So either of those changes should resolve the issue showing up in your error log, but as I said above, I'm afraid HTTPS support hasn't been tested in Cronicle at all, so I am sure there are other issues lurking.  If you can, please disable HTTPS and go with HTTP for now, until I can find the time to make HTTPS work properly across all features.  I'll add a separate GitHub issue to track this.

Thanks again for the bug report!

**Comment by mpatel-cpi** [comment]
Thanks for the quick response. I've been scratching my head around this issue for last 4-5 days, wish I had reported the issue earlier :) Seriously, thx for the helpful response again.

After setting https_force: 0, jobs finish fine now, but I still see the following issues. I'll try out the newer build and update you.

## When job is run on Cronicle master
- "Live Job Log Event" shows live logs
- After the job completes "Job Log Event" is still empty
- "Download Log" and "View Full Log" show the job logs correctly

## When job is run on Cronicle slave
- "Live Job Log Event" is empty
- After the job completes "Job Log Event" is still empty
- "Download Log" and "View Full Log" show the job logs correctly

**Comment by jhuckaby** [comment]
Yeah, I suspect there are going to be a number of HTTPS issues like this.  I was just trying to get you into a semi-working state, until I can devote some time to #27.

**Comment by danielhass** [comment]
Hey there, first of all thumbs up for the great Cronicle project! 

We're having nearly the same issue in our setup - but we don't use SSL at all.

# Environment
- Cronicle Version: 0.8.19
- NodeJS Version: v6.11.1

# Problem

Jobs that are running on the Master show the live log view and the log file in the end correctly. Jobs on Slave hosts report progress and status correctly but the log doesn't work:

- "Live Job Log Event" shows the following - but no live logs are displayed afterwards:
```
Log Watcher: Connecting to server: http://<X.X.X.X>:3012...
Log Watcher: Connected successfully!
```
- After the job completes "Job Log Event" shows: "(No log file found.)"
- "Download Log" and "View Full Log" show: "(No log file found.)"

We saw this Error.log message on the master:

```
[1533829549.11][2018-08-09 17:45:49][ideweiiss8160][Error][error][job][Failed to fetch job log file: http://<X.X.X.X>:3012/api/app/fetch_delete_job_log?path=%2Fopt%2Fdbmgmt%2Fapplication%2FCronicle%2Flogs%2Fjobs%2Fjjkmqiaos2c.log&auth=ebdd446072d9ee595bfc67365bf21b7f4dbc9493c1dd44428ea623e3a3db26bf: Error: ENOENT: no such file or directory, open '/opt/dbmgmt/application/Cronicle/logs/jobs/jjkmqiaos2c.log'][]
```

And this Cronicle.log message on the slave that is executing the job:

```
[1533829549.125][2018-08-09 17:45:49][<Hostname>][Cronicle][debug][4][Deleting log file: /opt/dbmgmt/application/Cronicle/logs/jobs/jjkmqiaos2c.log][]
```

@jhuckaby - any ideas on this one?

**Comment by danielhass** [comment]
@jhuckaby  - thanks for your fast reply. After moving the Cronicle installation to `/opt/cronicle` the log is displayed correctly. With the old setup we also had a symlink deployed like this `/opt/cronicle -> /opt/dbmgmt/application/Cronicle/` but this doesn't seem to help.

The live log does not work anyway but this is not that important for our setup..

**Comment by jhuckaby** [comment]
@danielhass Ah good, I'm glad that fixed at least some of the problem.  I have no idea why the live log isn't working, alas.
